### ISSUE TITLE
Error at screen curtain activation when Windows Magnifier is active

### ISSUE THREAD

**Issue by CyrilleB79** [nvaccess/nvda#10545]
### Steps to reproduce:

* Launch Windows Magnifier (Windows+PavNumPlus)
* Alt+Tab to Magnifier window
* Press "Display" button and select "Full screen" option
* Press "Options" button and check the "Invert colors" checkbox
* Open Vision panel in NVDA settings
* Uncheck the "Always show a warning" checkbox (probably does not matter but I did this way)
* Check the "Enable screen curtain" checkbox

### Actual behavior:

* Screen curtain is not enabled
* an error beep is heard: cf. [logWinMagnifier+ScreenCurtain.txt](https://github.com/nvaccess/nvda/files/3886970/logWinMagnifier%2BScreenCurtain.txt)
* a message box indicating"Could not load the Screen Curtain vision enhancement provider" appears.

### Expected behavior:

* Best choice: screen curtain is enabled and the screen becomes black
* Mid choice: screen curtain is enabled, but the screen is white due to color inversion of the magnifier
* Minimal choice: screen curtain cannot be activated, no error appears in the log and the message explains to the user why screen curtain cannot be activated

### Other tests performed

* Test 2: Windows Magnifier with normal colors + screen curtain: no error, the screen becomes black
* Test 3: Zoomtext 11 (normal colors) + screen curtain: no error, the screen becomes black
* Test 4: Zoomtext 11 (inverted colors) + screen curtain: no error, the screen becomes white
* Test 5:
  * Start Windows Magnifier with normal colors
  * Activate screen curtain: no error, screen black
  * Invert Windows Magnifier colors: no error, the screen remains black (so this seems possible...)

#### NVDA installed/portable/running from source:
From source

#### NVDA version:

Version : source-master-c750926
(last version)

#### Windows version:

Windows 10 1709

#### Name and version of other software in use when reproducing the issue:

For test 1: none.
For some other tests: Zoomtext 11

#### Does the issue still occur after restarting your PC?

Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.

No other stable version.
I have already seen this issue few time after the screen curtain was merged into alpha but did not bother to fill an issue at the time. I do not remember however the version of the alpha in which I have already seen the issue.

**Comment by leonardder** [comment]
I'm afraid the only thing we could do here is improve the UX. As windows Magnifier holds the API that we use to activate the screen curtain, we simply can't use the screen curtain when the magnifier is on.

**Comment by leonardder** [comment]
It should be pretty easy to at least add a warning to the NVDA user guide telling that the screen curtain can't be enabled when the windows magnifier is on. We could also hide the curtain option when the magnifier is on.

@CyrilleB79: could you also check what happens if the curtain is on and you activate the widnows magnifier? I"m afraid the curtain is killed without notice? Again, this is also a case that's pretty difficult to act upon.

**Comment by CyrilleB79** [comment]
> As windows Magnifier holds the API that we use to activate the screen curtain, we simply can't use the screen curtain when the magnifier is on.

Yes we can as described briefly in Test 2: Screen curtain can be activated when Windows Magnifier is active, provided that color inversion is not enabled at this time.

**Comment by CyrilleB79** [comment]
@leonardder I have made the test you suggest. Here are the results:

Test 6:

* Activate Windows Magnifier (Windows+PavNumPlus) and check in options that color inversion is active
* Unactivate Magnifier (Windows+Escape)
* Activate and unactivate Magnifier once again to check that color inversion is kept in memory between 2 Magnifier launch.
* Once Windows Magnifier is not active, start NVDA's screen curtain:
  Result: the screen becomes black
* Launch Windows Magnifier (Windows+PavNumPlus)
  Result: No error happens but nothing can be seen since the screen is black
* Disable NVDA's screen curtain
  Result: the screen is magnified but the colors are not inverted anymore!
* Quit Win Magnifier (Windows+Escape)
* Re-launch Windows Magnifier
  Result: Magnification OK and color inversion OK

Maybe Windows Magnifier programs a color transformation matrix that is overridden by NVDA's screen curtain color transformation matrix.

The easiest solution is to implement the minimal choice for expected behaviour, i.e. indicate to the user that screen curtain should not be used when Magnifier is active. However we should ideally also discourage the user to launch the Magnifier when screen curtain is on.

Another question is: Is screen curtain profile-dependant?
If yes, what will happen if the user is using the magnifier (inverted colors) and switch to an application wwhose profile asks to enable the magnifier? I did not yet made the test...

**Comment by CyrilleB79** [comment]
By the way, @leonardder, why did you put the CANTFIX label? Probably we cannot support screen curtain along with Windows Magnifier. However we can probably do some things to improve UX:

* Give the user a user-friendly message when he tries to activate the screen curtain while the Magnifier is active
* Avoid the error in the log
* Update user guide to warn the user not to use Magnifier along with screen curtain

**Comment by leonardder** [comment]
> By the way, @leonardder, why did you put the CANTFIX label?

Because we can't fix any conflicts between the magnifier and screen curtain. We can only describe them in the user guide.

Your tests really prove to be helpful, thanks a lot for this!

**Comment by CyrilleB79** [comment]
OK, I thought that CANTFIX label was a label to use when an issue was closed to indicate the closure reason. This incompatibility should however be indicated at least in user doc (and maybe in a message box when activating feature) before closing this issue.

**Comment by CyrilleB79** [comment]
Another idea: Would it be possible to use MagGetFullscreenColorEffect before the screen curtain activation in order to store the current color transformation? This would allow to restore the color transformation when screen curtain is deactivated.

**Comment by Adriani90** [comment]
@CyrilleB79 I see that a related PR has been merged. Did this improve it now?
If there are some conflicts with magnifier and screen curtain, I agree these should be documented in the user guide.

**Comment by CyrilleB79** [comment]
No, the merged PR (#10689) does not solve this issue, i.e. it does not allow to activate screen curtain when Magnifier is active. There does not seem to be a solution. So I think documenting this incompatibility would be enough to close this issue.

**Comment by CyrilleB79** [comment]
As explained in https://github.com/nvaccess/nvda/pull/16258#issuecomment-1978464234, to close this issue, not only the documentation should be updated to mention this limitation. In addition the screen curtain failure should be notified in some way to the user, also when triggering a profile (app switching) or when starting NVDA, what is not the case today.
